MT3-653 Display multiple tenants on Review Sign & Submit

Gweaton opened the pull request:
We've decided to display all sections for all tenants in one long list, as we haven't got any filtering components at the moment to allow housing officers to select a single tenant to view on the Review page.

This PR also refactors a couple of checks for `undefined` values when calling `.result` on values that are returned from the database to make them slightly easier to work with.

Gweaton committed:
Extract Review section into own component

This will allow us to render ReviewSections for each tenant

Gweaton committed:
Add Id and Residency review section

Gweaton committed:
Add Household review section

Gweaton committed:
Add Property Inspection review section

Gweaton committed:
Add Wellbeing support review section

Gweaton committed:
Render review sections for each present tenant

This also includes a couple of refactors:
- Rework tenant name & id fetching methods
- Check `.result` of values returned from database early, so we don't
need so many `undefined` checks later on in the file.

Gweaton committed:
Render Signature component for each present tenant

erbridge commented on components/ReviewSection.tsx:
```diff
@@ -0,0 +1,47 @@
+  const { heading, rows } = section || {};
+
+  return (
+    <React.Fragment key={heading}>
```
```suggestion
    <>
```

`key` doesn't mean anything on the components themselves, I don't think. (It does, but it's not relevant in this case.)

erbridge commented on components/review-sections/HouseholdReviewSection.tsx:
```diff
@@ -0,0 +1,16 @@
+    <ReviewSection
+      section={{
+        heading: "Household",
+        rows: [...useReviewSectionRows(Storage.ProcessContext, householdSteps)],
```
```suggestion
        rows: useReviewSectionRows(Storage.ProcessContext, householdSteps),
```

No need to destructure here.

erbridge replied:
Same for the other sections.

erbridge commented on pages/thc/[processRef]/review.tsx:
```diff
@@ -208,23 +154,19 @@ const ReviewPage: NextPage = () => {
-            )}
-          </React.Fragment>
-        ))}
+      {tenantsPresent.map(({ fullName, id }) => (
+        <React.Fragment key={id}>
+          <Heading level={HeadingLevels.H2}>{fullName}</Heading>
+          <IdAndResidencyReviewSection selectedTenantId={id} />
+          <HouseholdReviewSection />
+          <PropertyInspectionReviewSection />
+          <WellbeingSupportReviewSection />
+        </React.Fragment>
+      ))}
```
Ah I see what you've done. I think we probably want to only repeat the ID and residency check steps for each tenant, and to repeat it for all tenants, not just the present ones. We also ask for a signature from each present tenant, only. The housing officer would then start from the top of the page, confirming all of the details, get a tenant to sign, then repeat for the other present tenants.

If we do that, I think we want to pull the logic for iterating through the tenants into the `IdAndResidencyReviewSection`.

erbridge commented on pages/thc/[processRef]/review.tsx:
```diff
@@ -243,16 +185,21 @@ const ReviewPage: NextPage = () => {
-          setSignatures((sigs) => ({ ...sigs, [selectedTenantId]: value }));
-        }}
-      />
+      {tenantsPresent.map(({ fullName, id }) => (
+        <React.Fragment key={id}>
+          <Heading level={HeadingLevels.H3}>{fullName}</Heading>
+          <Signature
+            value={id && signatures[id]}
+            onChange={(value): void => {
+              if (!id) {
+                return;
+              }
+
+              setSignatures((sigs) => ({ ...sigs, [id]: value }));
+            }}
+          />
+        </React.Fragment>
+      ))}
```
👍

erbridge commented on pages/thc/[processRef]/review.tsx:
```diff
@@ -270,11 +217,13 @@ const ReviewPage: NextPage = () => {
+              tenantId,
+              signatures[tenantId] || ""
+            );
+          }
```
Ideally this would all be done in a transaction, but this is fine for now.

erbridge commented on components/ReviewSection.tsx:
```diff
@@ -0,0 +1,47 @@
+
+      {rows && rows.length ? (
+        <SummaryList
+          rows={(rows as unknown) as { key: string; value: string }[]}
```
Make sure you note the reason for this cast in the commit message if you haven't.

Gweaton replied:
I've actually removed this, and it doesn't seem to be breaking anything - unless I've misunderstood what the purpose of it was...

erbridge reviewed (commented):
I like the approach, and it makes the review page a lot simpler. A few comments, the biggest of which is clarifying what I meant by putting them all on top of each other.

Gweaton committed:
Pass multiple tenants into Id and Residency section

alongside their `isPresent` status.

This allows us to render the unique steps for each tenant on the Review
page, but only render the common content once.

erbridge commented on components/ReviewSection.tsx:
```diff
@@ -0,0 +1,45 @@
+import PropTypes from "../helpers/PropTypes";
+import { SectionRow } from "../helpers/useReviewSectionRows";
+
+interface Props {
+  section?: {
+    heading?: string;
+    rows?: SectionRow[];
+  };
+}
```
```suggestion
interface Props {
    heading?: string;
    rows?: SectionRow[];
}
```

This would be more idiomatic, I think.

erbridge commented on components/review-sections/IdAndResidencyReviewSection.tsx:
```diff
@@ -0,0 +1,83 @@
+interface Tenant {
+  fullName: string;
+  id: string;
+  isPresent: boolean;
```
```suggestion
  present: boolean;
```

We tend to prefer this style to be consistent with HTML booleans (e.g. `required`, `checked`).

erbridge commented on components/review-sections/IdAndResidencyReviewSection.tsx:
```diff
@@ -0,0 +1,83 @@
+        {isPresent ? "" : " (not present)"}
+      </Heading>
+      <SummaryList
+        rows={useReviewSectionRows(
```
Can we pull the hooks out of being inline to make them easier to find? Feels a bit weird to have it here, and makes conditional rendering difficult to add, with possibly confusing errors. I usually try to keep them at the top of the function, so any conditional logic won't get in the way.

This applies to all of the sections.

erbridge commented on components/review-sections/IdAndResidencyReviewSection.tsx:
```diff
@@ -0,0 +1,83 @@
+      <Heading level={HeadingLevels.H2}>
+        {"ID, residency, and tenant information"}
+      </Heading>
+      <SummaryList
+        rows={useReviewSectionRows(
+          Storage.ProcessContext,
+          idAndResidencyProcessSteps
+        )}
+      />
```
What does this render if the hook returns an empty array?

Gweaton replied:
This is the only thing that's left that's a little off - `presentForCheck` is the only `idAndResidencyProcessStep`, and is always coming back as an empty array for me. In the sub-component I'm rendering "Loading..." now if the ID steps come back as empty, but for some reason this is always empty...

erbridge commented on pages/thc/[processRef]/review.tsx:
```diff
@@ -117,46 +105,10 @@ const ReviewPage: NextPage = () => {
-    setSelectedTenantId(tenantIds[0]);
+  if (selectedTenantId !== presentTenantIds[0]) {
+    setSelectedTenantId(presentTenantIds[0]);
   }
```
I think this can disappear entirely now, along with the `useState` hook that created it?

erbridge commented on pages/thc/[processRef]/review.tsx:
```diff
@@ -188,7 +140,7 @@ const ReviewPage: NextPage = () => {
             {
               key: "Tenants",
-              value: tenantNames ? tenantNames.join(", ") : "Loading...",
+              value: allTenantNames ? allTenantNames.join(", ") : "Loading...",
```
```suggestion
              value: allTenantNames.length > 0 ? allTenantNames.join(", ") : "Loading...",
```

erbridge commented on pages/thc/[processRef]/review.tsx:
```diff
@@ -208,23 +160,14 @@ const ReviewPage: NextPage = () => {
       {tenantsPresent.length > 0 && (
-        <Paragraph>Present for check: {tenantsPresent.join(", ")}</Paragraph>
+        <Paragraph>
+          Present for check: {presentTenantNames.join(", ")}
```
```suggestion
          Present for check: {presentTenantNames.length > 0 > presentTenantNames.join(", ") : "Loading..."}
```

erbridge reviewed (commented):
Some suggestions. Nothing drastic.

Gweaton committed:
Remove `selectedTenantId` hook

We're now rendering all tenants on the review page, so this behaviour is
no longer needed

erbridge commented on components/review-sections/IdAndResidencyReviewSection.tsx:
```diff
@@ -0,0 +1,96 @@
+  return (
+    <>
+      <Heading level={HeadingLevels.H2}>
+        {"ID, residency, and tenant information"}
```
```suggestion
        ID, residency, and tenant information
```

Nitpick if you're back in these files.

erbridge commented on pages/thc/[processRef]/review.tsx:
```diff
@@ -207,24 +152,16 @@ const ReviewPage: NextPage = () => {
-            )}
-          </React.Fragment>
-        ))}
+      <Paragraph>
+        Present for check:{" "}
+        {presentTenantNames.length > 0
+          ? presentTenantNames.join(", ")
+          : "Loading..."}
+      </Paragraph>
```
I reckon this can be generated by the `IdAndResidencyReviewSection` now, so let's do that.

Gweaton replied:
As discussed, we're leaving this as the review section just has the IDs, which would duplicate the lookups for names (as we need to pass that data through to the ReviewSection anyway)
